src/main.zig

```diff
@@ -393,19 +486,66 @@ pub fn Deserializer(comptime ReaderType: type) type {
                 else => return error.MismatchingFormatType,
             };
 
-            data_type.* = try reader.readIntBig(i8);
+            var _data_type = try reader.readIntBig(i8);
+
             const buffer = try self.allocator.alloc(u8, len);
-            const actual_len = try reader.readAll(buffer);
+            errdefer self.allocator.free(buffer);
 
+            const actual_len = try reader.readAll(buffer);
             if (actual_len < len) return error.EndOfStream;
 
+            // Do not replace the value of input variable `data_type` unless we
+            // are certain this function call will not fail
+            data_type.* = _data_type;
+
             return buffer;
         }
 
-        /// Resets the internal buffer of `Self`. Calling any of the deserializing functions
-        /// after this will rewrite the buffer starting at index 0.
-        pub fn reset(self: *Self) void {
-            self.index = 0;
+        /// Recursively frees any value that was returned by a `deserialize` function (or any of it's specific variants)
+        pub fn free(self: *Self, ptr: anytype) void {
+            const T = @TypeOf(ptr);
+            comptime assert(trait.is(.Pointer)(T));
+
+            const C = comptime meta.Child(T);
+
+            if (comptime trait.hasFn("deinit")(C)) return C.deinit(ptr, self);
```

Comment:
Can you think of specific use cases where this is required that aren't solved by the below logic?
I'm not much of a fan of this hidden behavior (although you could argue the same of the `deserialize` and `serialize` methods)

Reply:
The use case was specifically that. Since the the serialization and deserialization allow for a custom method to be declared, for consistency I also think the freeing of the memory should be customizable. If we were to remove one, we ought to remove both I think. Specifically because in a custom `deserialize` method, the user could allocate memory himself, and so he would need some way of being able to free that memory too, or it would leak.

While I agree with the fact that his can feel a little bit like a hack (checking for the existence of a special named method at compile time) I kinda think there might be value to allowing certain structures to be serialized/deserialized (allocated)/deinit (freed) in a custom way, don't you agree?

Reply:
That does sound reasonable. Especially the use case where the user allocates memory within their custom `deserialize` method.